### PR TITLE
Feature temper likelihood - close #324

### PR DESCRIPTION
**Pull request by elbaraim**
reference here #324 
- [ ] I needed a model example using priors I could try to implement one

### CODE REVIEW

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/metropolis.py  1
- pypesto/sampling/result.py  5
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by yannikschaelte** [comment]
> I needed a model example using priors I could try to implement one

There is none, to my knowledge, or did @jvanhoefer implement one? Ideally via petab.

**Comment by yannikschaelte** [comment]
For the loggers in the notebooks, consider using https://github.com/ICB-DCM/pyPESTO/pull/398#event-3445713913 once merged.

File: pypesto/sampling/metropolis.py [hunk 1]
```diff
@@ -29,31 +31,41 @@ def default_options(cls):
...
     def initialize(self, problem: Problem, x0: np.ndarray):
         self.problem = problem
-        self.objective = problem.objective
-        self.objective.history = History()
+        self.neglogpost = problem.objective
+        self.neglogpost.history = History()
+        if problem.x_priors is None:
+            self.neglogprior = lambda x: -0.
+        else:
+            self.neglogprior = problem.x_priors
```
Comment (yannikschaelte) [inline pypesto/sampling/metropolis.py:39]:
```suggestion
        neglogprior = problem.x_priors
        if neglogprior is None:
            neglogprior = lambda x: -0.
        self.neglogprior = neglogprior
```

File: pypesto/sampling/metropolis.py [hunk 2]
```diff
@@ -29,31 +31,41 @@ def default_options(cls):
...
         self.trace_x = [x0]
-        self.trace_fval = [self.objective(x0)]
+        self.trace_neglogpost = [self.neglogpost(x0)]
+        self.trace_neglogprior = [self.neglogprior(x0)]
 
-    def sample(self, n_samples: int, beta: float = 1.):
+    def sample(self, n_samples: int, beta: float = 1.,
+               temper_lpost: bool = False):
```
Comment (yannikschaelte) [inline pypesto/sampling/metropolis.py:45]: can't add `temper_lpost` ad an argument to the sample function. please define it in `__init__`.

File: pypesto/sampling/metropolis.py [hunk 3]
```diff
@@ -29,31 +31,41 @@ def default_options(cls):
...
 
         show_progress = self.options['show_progress']
 
         # loop over iterations
         for _ in tqdm(range(int(n_samples)), disable=not show_progress):
             # perform step
-            x, llh = self._perform_step(x, llh, beta)
-
+            x, lpost, lprior = self._perform_step(x, lpost, lprior,
+                                                  beta, temper_lpost)
```
Comment (yannikschaelte) [inline pypesto/sampling/metropolis.py:57]: please use kwargs to avoid passing lpost, lprior the wrong way round.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

File: pypesto/sampling/metropolis.py (hunk 3 above)
Comment (elbaraim) [reply pypesto/sampling/metropolis.py]: I think this is addressed now, let me know whether this suffices

**Comment by jvanhoefer** [comment]
> > I needed a model example using priors I could try to implement one
> 
> There is none, to my knowledge, or did @jvanhoefer implement one? Ideally via petab.

nope, I didn't...
Maybe one could think of using the [Rosenbrock](https://en.wikipedia.org/wiki/Rosenbrock_function) example and interpret the (a-x)^2 term as Gaussian prior?

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/metropolis.py  1
- pypesto/sampling/result.py  5
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/metropolis.py  1
- pypesto/sampling/result.py  5
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

File: pypesto/sampling/metropolis.py (hunk 2 above)
Comment (yannikschaelte) [reply pypesto/sampling/metropolis.py:45]: TODO

File: pypesto/sampling/metropolis.py [hunk 4]
```diff
@@ -29,85 +31,118 @@ def default_options(cls):
...
         # loop over iterations
         for _ in tqdm(range(int(n_samples)), disable=not show_progress):
+            kwargs = {'lpost': lpost,
+                      'lprior': lprior}
             # perform step
-            x, llh = self._perform_step(x, llh, beta)
-
+            x, lpost, lprior = self._perform_step(x, beta,
+                                                  temper_lpost,
+                                                  **kwargs)
```
Comment (yannikschaelte) [inline pypesto/sampling/metropolis.py:60]: sorry, I meant
```suggestion
            x, lpost, lprior = self._perform_step(
                x=x, lpost=lpost, lprior=lprior, beta=beta,
                temper_lpost=temper_lpost)
```

Comment (elbaraim) [reply pypesto/sampling/metropolis.py:60]: i do not get it... you suggested to use kwargs to not to mess up with the order of the inputs??

Comment (yannikschaelte) [reply pypesto/sampling/metropolis.py:60]: yes, which the version I posted exactly does (it uses kwargs ...) ...

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/metropolis.py  1
- pypesto/sampling/result.py  5
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by jvanhoefer** [comment]
Maybe just a minor comment, but in 

https://github.com/ICB-DCM/pyPESTO/blob/e2b0ae0f6b445fbb909038d13ee1d0e03a29eb9b/pypesto/sampling/parallel_tempering.py#L23

shall we maybe use 
`if (betas is None) == (n_chains is None):`
I think it is more clean to avoid this type casting from bool to int ?

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

File: pypesto/sampling/parallel_tempering.py [hunk 5]
```diff
@@ -61,7 +64,7 @@ def initialize(self,
     def sample(
             self, n_samples: int, beta: float = 1.):
         # loop over iterations
-        for i_sample in tqdm(range(int(n_samples))):
+        for i_sample in tqdm(range(int(n_samples))):  # TODO test
```
Comment (yannikschaelte) [inline pypesto/sampling/parallel_tempering.py:67]: TODO

File: pypesto/sampling/pymc3.py [hunk 6]
```diff
@@ -104,19 +103,21 @@ def get_samples(self) -> McmcPtResult:
...
+        if trace_x.shape[0] != trace_neglogpost.shape[0] \
+                or trace_x.shape[1] != trace_neglogpost.shape[1] \
                 or trace_x.shape[2] != len(self.problem.x_names):
             raise ValueError("Trace dimensions are inconsistent")
 
         return McmcPtResult(
             trace_x=np.array(trace_x),
-            trace_fval=np.array(trace_fval),
+            trace_neglogpost=np.array(trace_neglogpost),
+            trace_neglogprior=np.full(trace_neglogpost.shape, np.nan),
```
Comment (yannikschaelte) [inline pypesto/sampling/pymc3.py:121]: yes, can't do anything about this atm. maybe something for @lcontento to be aware of.

File: pypesto/sampling/sampler.py [hunk 7]
```diff
@@ -126,9 +129,14 @@ def set_last_sample(self, sample: InternalSample):
...
-    def make_internal(self):
+    def make_internal(self, temper_lpost: bool):
         """
         This function can be called by parallel tempering samplers during
         initialization to allow the inner samplers to adjust to them
         being used as inner samplers. Default: Do nothing.
+
+        Parameters
+        ----------
+        tamper_lpost:
```
Comment (yannikschaelte) [inline pypesto/sampling/sampler.py:140]:
```suggestion
        temper_lpost:
```
correcting myself :see_no_evil:

File: test/test_sampling.py [hunk 8]
```diff
@@ -128,6 +145,73 @@ def test_ground_truth():
...
+    rvs1 = norm.rvs(size=5000, loc=-1., scale=np.sqrt(0.7))
+    rvs2 = norm.rvs(size=5001, loc=100., scale=np.sqrt(0.8))
+
+    # test
+    statistic, pval = ks_2samp(np.concatenate([rvs1, rvs2]),
+                               samples)
+    print(statistic, pval)
+    assert statistic < 0.1
+
+    sampler = pypesto.AdaptiveMetropolisSampler()
```
Comment (yannikschaelte) [inline test/test_sampling.py:174]: describe what the 2 tests below here are supposed to do.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/metropolis.py  1
- pypesto/sampling/result.py  5
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

File: pypesto/sampling/sampler.py [hunk 9]
```diff
@@ -32,7 +32,7 @@ def initialize(self,
 
     @abc.abstractmethod
     def sample(
-            self, n_samples: int, beta: float = 1.
+            self, n_samples: int, beta: float = 1., temper_lpost: bool = False
```
Comment (jvanhoefer) [inline pypesto/sampling/sampler.py]: Code format: No line break before `self`. One line per argument...

Comment (yannikschaelte) [reply pypesto/sampling/sampler.py]: it depends -- sometimes we do one argument per line, but if it fits in one or two lines all-together, we put all bunched. always trying to make it as readable as possible -- is it more important to have a concise code (and not have to look up 100 lines of arguments) or can't we find an argument if it's all just lined up.

**Comment by yannikschaelte** [comment]
> We still need to add doc strings to quite some of the functions/classes that were implemented in the sampling. Besides that, it looks fine for me :)

agreed

File: pypesto/sampling/sampler.py (hunk 9 above)
Comment (elbaraim) [reply pypesto/sampling/sampler.py]: so this is fine as it is?

File: pypesto/sampling/parallel_tempering.py (hunk 5 above)
Comment (elbaraim) [reply pypesto/sampling/parallel_tempering.py:67]: FYI: This was basically to also enable the switch-off of this progress bar.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/result.py  5
- pypesto/sampling/metropolis.py  1
         

Clones added
============
- test/test_sampling.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/sampling/metropolis.py  1
- pypesto/sampling/result.py  5
         

Clones added
============
- test/test_sampling.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5698685&bid=18653923)

**Pull request merged**